plugins/emoji/plugin.js

```diff
@@ -11,132 +11,96 @@
-			// jscs:enable
+			var emojiListUrl = editor.config.emoji_emojiListUrl || 'plugins/emoji/emoji.json';
 
-			if ( !editor._.emojiList ) {
-				editor._.emojiList = defaultEmojiList;
-			}
-			var emoji = editor._.emojiList,
-				forbiddenScope = editor.config.emoji_blacklistedElements || [ 'pre', 'code' ],
-				charactersToStart = editor.config.emoji_minChars || 2;
-
-			editor.on( 'instanceReady', function() {
-				new CKEDITOR.plugins.autocomplete( editor, {
-					textTestCallback: getTextTestCallback(),
-					dataCallback: dataCallback,
-					itemTemplate: '<li data-id="{id}" class="cke_emoji_suggestion_item">{symbol} {id}</li>',
-					outputTemplate: '{symbol}'
-				} );
-			} );
-
-			editor.on( 'toHtml', function( evt ) {
-				var sel = evt.editor.getSelection();
-				// We want to prevent embedding emoji inside wrong context, e.g. paste :emoji: inside <pre>
-				if ( sel && !isEmojiAllowed( sel.getRanges()[ 0 ] ) ) {
+			CKEDITOR.ajax.load( CKEDITOR.getUrl( emojiListUrl ), function( data ) {
+				if ( data === null ) {
 					return;
 				}
 
-				var filter = new CKEDITOR.htmlParser.filter( {
-					text: function( value, element ) {
-						var preventEmojiConversion = element.getAscendant( hasForbiddenParent );
-						if ( preventEmojiConversion ) {
-							return value;
-						}
-						var hits = value.match( /:[a-zA-Z_-]+?:/g );
-						if ( hits ) {
-							for ( var i = 0; i < hits.length; i++ ) {
-								var item = emoji.find( function( element ) {
-									return element.id === hits[ i ];
-								} );
-								if ( item ) {
-									value = value.replace( item.id, item.symbol );
-								}
-							}
-						}
-						return value;
-					}
-				} );
-				filter.applyTo( evt.data.dataValue );
-			} );
+				if ( editor._.emoji === undefined ) {
+					editor._.emoji = {};
+				}
 
-			function getTextTestCallback() {
-				return function( range ) {
-					if ( !range.collapsed || !isEmojiAllowed( range ) ) {
-						return null;
-					}
-					return CKEDITOR.plugins.textMatch.match( range, matchCallback );
-				};
-			}
+				if ( editor._.emoji.list === undefined ) {
+					editor._.emoji.list = JSON.parse( data );
+				}
 
-			function matchCallback( text, offset ) {
-				var left = text.slice( 0, offset ),
-					match = left.match( new RegExp( ':\\S{' + charactersToStart + '}\\S*$' ) );
+				var emojiList = editor._.emoji.list,
+					charactersToStart = editor.config.emoji_minChars === undefined ? 2 : editor.config.emoji_minChars;
 
-				if ( !match ) {
-					return null;
+				if ( editor.status !== 'ready' ) {
```

Comment:
Is this `if` needed now, when substitution is removed and initialisation just creates autocomplete instance?

Reply:
yes, it's required because of this bug: https://github.com/ckeditor/ckeditor-dev/issues/2114

Inlining this function cause random error, where plugin might be not initialised during loading stage (depend how fast json will be loaded). Error was clearly visible in Firefox browser.